// @filmgardi/videojs-next-episode/src/NextEpisodeButton.js
import videojs from "video.js";
import document from "global/document";
import window from "global/window";
import NextEpisodeDialog from "./NextEpisodeDialog";

const Button = videojs.getComponent("Button");
const Component = videojs.getComponent("Component");

class NextEpisodeButton extends Button {
  /**
   * Creates an instance of this class.
   *
   * @param {Player} player
   *        The `Player` that this class should be attached to.
   *
   * @param {Object} [options]
   *        The key/value store of player options.
   */
  constructor(player, options) {
    super(player, options);
    this.nextEpisodeDialog = new NextEpisodeDialog(player, options);
  }

  /**
   * Builds the default DOM `className`.
   *
   * @return {string}
   *         The DOM `className` for this object.
   */
  buildCSSClass() {
    return `vjs-next-episode-control ${super.buildCSSClass()}`;
  }

  /**
   * This gets called when an `FullscreenToggle` is "clicked". See
   * {@link ClickableComponent} for more detailed information on what a click can be.
   *
   * @param {EventTarget~Event} [event]
   *        The `keydown`, `tap`, or `click` event that caused this function to be
   *        called.
   *
   * @listens tap
   * @listens click
   */
  handleClick(event) {
    if (window.innerWidth <= 800) {
      const voteCard = document.getElementsByClassName("vjs-vote-card")[0];
      if (
        voteCard &&
        voteCard.classList.value.indexOf("vjs-hidden") === -1
      ) {
        voteCard.parentNode.removeChild(voteCard);
      }
    }
    this.nextEpisodeDialog.showDialog();
  }
}
NextEpisodeButton.prototype.controlText_ = "NextEpisode";
NextEpisodeButton.prototype.name = () => "NextEpisodeButton";
Component.registerComponent("NextEpisodeButton", NextEpisodeButton);
export default NextEpisodeButton;
